dataplane/internal/engine/engine.go

```diff
@@ -233,45 +282,249 @@ func (e *Engine) AddLayer3PuntRule(ctx context.Context, portName string, ip []by
+		}
+		ip = ipNet.IP.To4()
+		mask = ipNet.Mask
+		isIPv4 = true
+		if ip == nil {
+			ip = ipNet.IP.To16()
+			mask = ipNet.Mask
+			isIPv4 = false
+		}
+	case *dpb.RoutePrefix_Mask:
+		ip = pre.Mask.Addr
+		mask = pre.Mask.Mask
+		switch len(ip) {
+		case net.IPv4len:
+			isIPv4 = true
+		case net.IPv6len:
+			isIPv4 = false
+		default:
+			return ip, mask, isIPv4, vrf, fmt.Errorf("invalid ip addr length")
+		}
+	default:
+		return ip, mask, isIPv4, vrf, fmt.Errorf("invalid prefix type")
+	}
+	return ip, mask, isIPv4, vrf, nil
+}
+
+// addNextHopList creates all the next hops from the message, then create a next hop group if there are multiple next hops.
+func (e *Engine) addNextHopList(ctx context.Context, nhg *dpb.NextHopList) ([]*fwdpb.ActionDesc, error) {
+	if len(nhg.GetHops()) == 1 {
+		nhID := e.nextNHID.Add(1)
+		if err := e.addNextHop(ctx, nhID, nhg.Hops[0]); err != nil {
+			return nil, err
+		}
+		return []*fwdpb.ActionDesc{
+			fwdbuilder.Action(fwdbuilder.UpdateAction(fwdpb.UpdateType_UPDATE_TYPE_SET, fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_ID).WithUint64Value(nhID)).Build(),
+			fwdbuilder.Action(fwdbuilder.LookupAction(nhTable)).Build(),
+		}, nil
+	}
+
+	idList := &dpb.NextHopIDList{}
+	for _, hop := range nhg.GetHops() {
+		nhID := e.nextNHID.Add(1)
+		if err := e.addNextHop(ctx, nhID, hop); err != nil {
+			return nil, err
+		}
+		idList.Hops = append(idList.Hops, nhID)
+		idList.Weights = append(idList.Weights, hop.Weight)
+	}
+	nhgID := e.nextNHGID.Add(1)
+	if err := e.addNextHopGroupIDList(ctx, nhgID, idList); err != nil {
+		return nil, err
+	}
+	return []*fwdpb.ActionDesc{
+		fwdbuilder.Action(fwdbuilder.UpdateAction(fwdpb.UpdateType_UPDATE_TYPE_SET, fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_GROUP_ID).WithUint64Value(nhgID)).Build(),
+		fwdbuilder.Action(fwdbuilder.LookupAction(nhgTable)).Build(),
+	}, nil
+}
+
+// addNextHopGroupIDList adds an entry to the next hop group table.
+func (e *Engine) addNextHopGroupIDList(ctx context.Context, id uint64, nhg *dpb.NextHopIDList) error {
+	var actLists []*fwdpb.ActionList
+	for i, nh := range nhg.GetHops() {
+		actLists = append(actLists, &fwdpb.ActionList{
+			Weight: nhg.Weights[i],
+			Actions: []*fwdpb.ActionDesc{{
+				ActionType: fwdpb.ActionType_ACTION_TYPE_UPDATE,
+				Action: &fwdpb.ActionDesc_Update{
+					Update: &fwdpb.UpdateActionDesc{
+						FieldId: &fwdpb.PacketFieldId{
+							Field: &fwdpb.PacketField{
+								FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_ID,
+							},
+						},
+						Type:  fwdpb.UpdateType_UPDATE_TYPE_SET,
+						Value: binary.BigEndian.AppendUint64(nil, nh),
+					},
+				},
+			}},
+		})
+	}
+	actions := []*fwdpb.ActionDesc{{
+		ActionType: fwdpb.ActionType_ACTION_TYPE_SELECT_ACTION_LIST,
+		Action: &fwdpb.ActionDesc_Select{
+			Select: &fwdpb.SelectActionListActionDesc{
+				SelectAlgorithm: fwdpb.SelectActionListActionDesc_SELECT_ALGORITHM_CRC32, // TODO: should algo + hash be configurable?
```

Comment:
I think the comment got lost in the edit -- annotate that this is for ECMP/WCMP.